### PR TITLE
Reduce the duplicate code in IK and TSID

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR tries to reduce the duplicate code in IK and TSID and implements what has been discussed in #363.

I implemented the `ILinearTaskSolver` interface from which TSID and IK inherit. I also implemented `TSIDLinearTask` and `IKLinearTask` that are two specializations of `LinearTask`. Each IK task will inherit from `IKLinearTask` while each TSID task from `TSIDLinearTask`.  

No actions are required by the user to adapt the consumer libraries to this new implementation. 

cc @traversaro @isorrentino @S-Dafarra @paolo-viceconte

### CODE REVIEW

**Review by S-Dafarra — COMMENTED** [review]
What is your plan for ``IKLinearTask`` and ``TSIDLinearTask``? At the moment they simply inherit from ``LinearTask``, so I was wondering why not using ``LinearTask`` directly. Are you planning to add new methods in the future?

**Comment by GiulioRomualdi** [comment]
The idea is to avoid that someone uses a IK task in TSID and viceversa

**Comment by S-Dafarra** [comment]
> The idea is to avoid that someone uses a IK task in TSID and viceversa

Ah, I see. I also see why you used the template then, nice!

File: src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * ILinearTaskSolver describes the interface for solving problem related to LinearTask class. Please
+ * check IntegrationBasedIK for further details.
+ */
+template <class _Task, class _State> class ILinearTaskSolver : public Source<_State>
+{
+    static_assert(std::is_base_of_v<LinearTask, _Task>,
+                  "The _Task template argument of ILinearTaskSolver must inherits from "
+                  "LinearTask");
```
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply src/System/include/BipedalLocomotion/System/ILinearTaskSolver.h:38]: In this way, the solvers for the IK and the TSID are not allowed to use each other tasks.

**Review by S-Dafarra — APPROVED** [review]

**Comment by S-Dafarra** [comment]
The ``macOS`` job failed with error
```
2021-07-12T11:15:39.5015400Z Can't find any online and idle self-hosted or hosted runner in the current repository, account/organization that matches the required labels: 'macOS-latest'
2021-07-12T11:15:39.5015541Z Found online and busy hosted runner(s) in the current repository's organization account that matches the required labels: 'macOS-latest'. Hit concurrency limits on the hosted runners. Waiting for one of them to get assigned for this job.
2021-07-12T11:15:39.5015572Z Waiting for a hosted runner in 'organization' to pick this job...
```

Maybe it should be ``macos-latest`` (with the lowercase ``o``) in https://github.com/dic-iit/bipedal-locomotion-framework/blob/master/.github/workflows/ci.yml#L55 instead?

**Pull request merged**